packages/theme/src/cli/utilities/theme-uploader.ts

```diff
@@ -0,0 +1,86 @@
+import {mountThemeFileSystem, partitionThemeFiles} from './theme-fs.js'
+import {rejectLiquidChecksums} from './asset-checksum.js'
+import {AdminSession} from '@shopify/cli-kit/node/session'
+import {Theme, ThemeFileSystem} from '@shopify/cli-kit/node/themes/types'
+import {deleteThemeAsset, fetchChecksums} from '@shopify/cli-kit/node/themes/api'
+import {fileSize} from '@shopify/cli-kit/node/fs'
+
+interface UploadOptions {
+  path: string
+  nodelete?: boolean
+}
+
+// Limits for Bulk Requests
+const MAX_BATCH_FILE_COUNT = 10
+// 100KB
+const MAX_BATCH_BYTESIZE = 102400
+
+export async function uploadTheme(theme: Theme, session: AdminSession, options: UploadOptions) {
+  const localThemeFileSystem = await mountThemeFileSystem(options.path)
+  if (!options.nodelete) {
+    await deleteFiles(theme, session, localThemeFileSystem)
+  }
+
+  await uploadFiles(theme, session, localThemeFileSystem)
+}
+
+async function deleteFiles(theme: Theme, session: AdminSession, localThemeFileSystem: ThemeFileSystem) {
+  const remoteChecksums = await fetchChecksums(theme.id, session)
+  const filteredRemoteChecksums = rejectLiquidChecksums(remoteChecksums)
+
+  const filesPresentRemotely: string[] = filteredRemoteChecksums.map((checksum) => checksum.key)
+  const filesPresentLocally: Set<string> = new Set(localThemeFileSystem.files.keys())
+  const filesToDelete = filesPresentRemotely.filter((key) => !filesPresentLocally.has(key))
+
+  const {jsonFiles, liquidFiles, configFiles, staticAssetFiles} = partitionThemeFiles(filesToDelete)
+  const otherFiles = [...liquidFiles, ...configFiles, ...staticAssetFiles]
+
+  await Promise.all(otherFiles.map((file) => deleteThemeAsset(theme.id, file, session)))
+  await Promise.all(jsonFiles.map((file) => deleteThemeAsset(theme.id, file, session)))
+}
+
+async function uploadFiles(theme: Theme, session: AdminSession, localThemeFileSystem: ThemeFileSystem) {
+  const {jsonFiles, liquidFiles, configFiles, staticAssetFiles} = partitionThemeFiles([
+    ...localThemeFileSystem.files.keys(),
+  ])
+  await batchUpload(liquidFiles)
+  await batchUpload(jsonFiles)
+  await batchUpload(configFiles)
+  await batchUpload(staticAssetFiles)
+}
+
+async function batchUpload(files: string[]) {
+  await createBatches(files)
+}
+
+async function createBatches(files: string[]): Promise<string[][]> {
```

Comment:
Just want to confirm we don't need to handle threading / locking since we're in TS land